README.md
[![ci](https://github.com/nicovandenhooff/reddit-data-collector/actions/workflows/ci.yml/badge.svg)](https://github.com/nicovandenhooff/reddit-data-collector/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/nicovandenhooff/reddit-data-collector/branch/main/graph/badge.svg?token=G6KM5UCJ8N)](https://codecov.io/gh/nicovandenhooff/reddit-data-collector)
![license](https://img.shields.io/github/license/nicovandenhooff/reddit-data-collector)
![PyPI](https://img.shields.io/pypi/v/reddit-data-collector)

# Reddit Data Collector

Reddit Data Collector is a Python package that allows a user to collect post and comment data from Reddit. It is built on top of the Python module [PRAW](https://praw.readthedocs.io/en/stable/), which stands for "The Python Reddit API Wrapper". It aims to make it very simple for a user to collect data from Reddit for further analysis (e.g. Natural Language Processing), without having to learn the inner workings of PRAW or the Reddit API.

It is currently maintained by [Nico Van den Hooff](https://www.nicovandenhooff.com/).

## Installation

### Dependencies

Reddit Data Collector requires Python and:

- pandas (>=1.3.5)
- praw (>=7.5.0)
- tqdm (>=4.62.3)

### User installation

The recommended way to install Reddit Data Collector is using `pip`:

```shell
pip install reddit-data-collector
```

## How to Use Reddit Data Collector

Please see the [examples directory](https://github.com/nicovandenhooff/reddit-data-collector/tree/main/examples) for step by step instructions on how to use Reddit Data Collector.

## Development

### Important links

- Official source code repo: <https://github.com/nicovandenhooff/reddit-data-collector>
- Downloaded releases: <https://pypi.org/project/reddit-data-collector/>
- Issue tracker: <https://github.com/nicovandenhooff/reddit-data-collector/issues>

### Source code

You can check the latest sources with the command:

```shell
git clone https://github.com/nicovandenhooff/reddit-data-collector.git
```

### Contributing

To learn more about making a contribution to Reddit Data Collector, please see the [contributing file](https://github.com/nicovandenhooff/reddit-data-collector/blob/main/CONTRIBUTING.md).

#### Potential Ideas for Contribution

- Add ability to collect images from Reddit posts that contain them.
- Add author information to post and comment data, currently the Reddit API is inconsistent with suspended and deleted author data, so this functionality has not been built in yet.
- Add plotting module that creates useful visualizations of the data that has been collected
- Add preprocessing module that cleans up the posts and/or comment data collected

### Testing

After installation, you can launch the test suite, which is contained in the `tests/tests.py`.  Note that you will have to have `pytest` >= 6.2.5 and `pytest-cov` >= 3.0.0 installed.  You can launch the test suite by following these steps from the projects root directory:

1. Open up `tests.py` with the following command:

```bash
open tests/tests.py
```

Comment out lines 24 to 30.  Change the values in `DataCollector()` in line 32 to your Reddit credentials.

2. Run the following command to run the tests:

```shell
pytest tests/tests.py
```

3. If desired, run the following command to show test coverage:

```shell
pytest --cov=src tests/tests.py
```

## Project History

The project was started in January 2022 by Nico Van den Hooff as a side project while he was completing the UBC Master of Data Science Project.  Nico wanted to obtain a sample of posts and comments from Reddit, but noticed that while PRAW existed and provided seamless access to Reddit's API, there was no package available that allowed for a simple method to collect this data.

#### Inspiration

Certain sections of this README file was inspired by the [scikit-learn README](https://github.com/scikit-learn/scikit-learn/blob/main/README.rst).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2022-03-13

### Changed

- Changed `get_data` in `reddit_data_collector.py` to return pandas `DataFrame` by default
- Updated tests for the above

## [1.0.2] - 2022-01-14

### Fixed

- Updated `_check_subreddit_exists` in `reddit_data_collector.py` to check both names as `.lower()`
- Updated tests for the above

### Changed

- Updated `README` to include instructions on  coverage tests

## [1.0.1] - 2022-01-12

### Fixed

- Spelling error of `separate` argument in `to_pandas` function of `reddit_data_collector.io.py`, previously it was spelt like `seperate`

### Changed

- Update example use and move to `/examples`
- Update PyPi link in docs to working link
- Add new potential ideas for contribution

## [1.0.0] - 2022-01-07

The first release of Reddit Data Collector.


src/reddit_data_collector/io.py
"""
The module `reddit_data_collector.io` includes useful input/output functions
that assist in converting and/or saving data collected from Reddit with
Reddit Data Collector.
"""

import pandas as pd
from .exceptions import ColumnNameError


def to_pandas(subreddit_data, separate=False):
    """Convert raw post or comment data collected to a pandas `DataFrame`.

    Parameters
    ----------
    subreddit_data : dict
        Raw post or comment data collected with the `DataCollector.get_data`
        method.

    separate : bool, default=False
        Whether or not to return a separate pandas `DataFrame` for the
        data of each subreddit.

    Returns
    -------
    df or dfs : pd.DataFrame or dict
        If separate is `False`, returns a pandas `DataFrame` containing
        the post or comment data.

        If separate is `True` returns a Python dictionary containing
        a pandas `DataFrame` for each subreddit that existed in the
        post or comment data.  The dctionary keys are the subreddits
        names.  The dictionary values are pandas `DataFrame`s of post
        or comment data.

    See Also
    --------
    reddit_data_collector.reddit_data_collector.DataCollector
        Class that performs the data collection from Reddit.

    reddit_data_collector.io.update_data
        Update a `.csv` file containing existing post or comment
        data collected with new data collected with `DataCollector`.
    """
    dfs = dict()

    for subreddit, data in subreddit_data.items():
        dfs[subreddit] = pd.DataFrame(data)

    if separate:
        return dfs
    else:
        return pd.concat(dfs.values(), ignore_index=True)


def update_data(csv_path, df, key="id", sort="subreddit_name", save=True):
    """Update a `.csv` file containing post or comment data with new data.

    The main purpose of this method is to allow a user to update a `.csv`
    file that contains historical data that they collected with Reddit Data
    Collector with new data collected.  The default method settings ensure
    that duplicated post or comment data, if any, is not saved to the `.csv`
    file.  In other words only one copy of each post or comment is kept in
    the combined data.

    If the `save` parameter is set to `True` then the method will automatically
    overwrite the existing `.csv` file.  Otherwise it will just return the
    combined data to the user as a pandas `DataFrame` for which they can
    then save with the pandas `DataFrame.to_csv` method when desired.

    Parameters
    ----------
    csv_path : str
        The file path to the existing `.csv` file.

    df : pandas DataFrame
        A pandas `DataFrame` containing the new data collected.  It is
        recommended that this `DataFrame` comes from the output of the
        `to_pandas` method in `reddit_data_collector.io`.

    key : str, default="id"
        The key to remove duplicate data on.  Default is the post or
        comment `id` as set by Reddit.  It is not recommended to set
        this parameter manually.  However, it is included as a parameter
        in case for some reason duplicate data is desired.

    sort : str, default="subreddit_name"
        How to sort the new data.  By default sorts the data by subreddit.
        This is purely aesthetic and has no impact on the data itself.

    save : bool, default=True
        Whether or not to automatically overwrite the existing `.csv`
        file with the new data.

    Returns
    -------
    new_df : pandas DataFrame
        A pandas `DataFrame` containing the newly combined post or comment
        data.

    Raises
    ------
    ColumnNameError
        If the update is attempted with two pandas `DataFrame`s that have
        different column names.

    See Also
    --------
    reddit_data_collector.reddit_data_collector.DataCollector
        Class that performs the data collection from Reddit.

    reddit_data_collector.io.to_pandas
        Used to convert raw `posts` or `comments` collected with
        `DataCollector` to a pandas `DataFrame`.

    Examples
    --------
    >>> import reddit_data_collector as rdc
    >>> # create instance of DataCollector
    >>> data_collector = rdc.DataCollector(
    ...     "<your_client_id>",
    ...     "<your_client_secret>",
    ...     "mac:script:v1.0 (by u/FakeRedditUser)",
    ...     "FakeRedditUser",
    ...     "FakePassword"
    ... )
    >>> # collect some data from Reddit
    >>> subreddits = ["pics", "funny"]
    >>> post_filter = "hot"
    >>> comment_data = True
    >>> replies_data = True
    >>> posts, comments = data_collector(
    ...     subreddits=subreddits,
    ...     post_filter=post_filter,
    ...     comment_data=comment_data,
    ...     replies_data=replies_data
    ... )
    >>> # update existing .csv file
    >>> new_posts_df = rdc.update_data("post_data.csv", posts_df)
    >>> new_comments_df = rdc.update_data("comment_data.csv", comments_df)
    """

    if not set(pd.read_csv(csv_path).columns) == set(df.columns):
        raise ColumnNameError("Both data sets must have the same features")

    old_df = pd.read_csv(csv_path)

    new_df = (
        pd.concat([old_df, df], ignore_index=True)
        .drop_duplicates(subset=[key], ignore_index=True)
        .sort_values(sort, ignore_index=True)
    )

    if save:
        new_df.to_csv(csv_path, index=False)

    return new_df


src/reddit_data_collector/exceptions.py
"""
The module `reddit_data_collector.exceptions` includes all custom error classes
used across Reddit Data Collector.
"""


class SubredditError(Exception):
    """Exception class raised if invalid subreddit is used.

    Examples
    --------
    >>> from reddit_data_collector.exceptions import SubredditError
    >>> try:
    ...     data_collector.get_data(subreddits="1nv4ald")
    ... except SubredditError as e:
    ...     print(repr(e))
    SubredditError('r/1nv4ald does not exist')
    """

    pass


class FilterError(Exception):
    """Exception class raised if an invalid post or top post filter is used.

    Examples
    --------
    >>> from reddit_data_collector.exceptions import FilterError
    >>> try:
    ...     data_collector.get_data(subreddits="funny", post_filter="any")
    ... except FilterError as e:
    ...     print(repr(e))
    FilterError('Invalid post_filter used: any')
    >>> try:
    ...     data_collector.get_data(
    ...        subreddits="funny",
    ...        post_filter="top",
    ...        top_post_filter="now"
    ...     )
    ... except FilterError as e:
    ...     print(repr(e))
    FilterError('Invalid top_post_filter used: now')
    """

    pass


class ColumnNameError(Exception):
    """Exception class used if data update is attempted with mismatched columns.

    Examples
    --------
    >>> import pandas as pd
    >>> from reddit_data_collector.exceptions import ColumnNameError
    >>> csv_path = "example.csv"
    >>> # create and save first DataFrame
    >>> df = pd.DataFrame(data=[[1, 2], [3, 4]], columns=["a", "b"])
    >>> df.to_csv(path, index=False)
    >>> # create second DataFrame
    >>> df2 = pd.DataFrame(data=[[5, 6], [7, 8]], columns=["c", "d"])
    >>> try:
    ...    rdc.update_data(csv_path, df2)
    ... except ColumnNameError as e:
    ...    print(repr(e))
    ColumnNameError('Both data sets must have the same features')
    """

    pass


src/reddit_data_collector/reddit_data_collector.py
"""
Reddit data collector module for Python
=======================================

Reddit Data Collector is a Python module that allows a user to collect post
and comment data from Reddit.  It is built on top of the Python module PRAW,
which stands for "The Python Reddit API Wrapper".  It aims to make it very
simple for a user to collect data from Reddit, without having to learn the
inner workings of PRAW or the Reddit API.

Reddit Data Collector provides the ability to collect a sample of post data 
and comment data from Reddit by simply providing the subreddit names 
that you wish to collect data from.  By default, the data is returned in
the form of a Pandas DataFrame.  There is also functionality to seamlessly 
update an existing .csv file that contains some reddit data collected with 
the module in the past, with some new data that been collected.

The functionalities above should allow a user to build a good sample of Reddit 
data over a period of time.  For example, a user could collect a sample from 
Reddit each day for a month, and each time a sample is collected add it to a 
universal `.csv` file in order to compile the data. At the end of the month, 
the user would have a months worth data sampled from Reddit, for which they 
could now analyze further (e.g. with Natural Language Processing).

The creation of Reddit Data Collector was inspired since the Reddit API does
not allow you to perform one large scrape of historical data (e.g. collect 
all posts from subreddit X from January 2015 to December 2020).  Therefore
this module provides an alternative option to build a data set by collecting
data over time.
"""

import praw

from tqdm import tqdm
from .io import to_pandas
from .exceptions import SubredditError, FilterError


class DataCollector:
    """Object that performs data collection from Reddit.

    Once a `DataCollector` object is instantiated, you simply need to pass the subreddit
    name(s) that you desire to collect data from to the method `get_data`, and the data
    collection will be performed.

    Please see the Reddit's "First Step Guide" which describes how to obtain the
    `client_id` and `client_secret` parameters below:
    https://github.com/reddit-archive/reddit/wiki/OAuth2-Quick-Start-Example#first-steps

    Important: If you instantiate `DataCollector` without a Reddit username and password,
    it will have read only access to the reddit API, which is limited to 30 requests
    per minute.  However, if you do provide a Reddit username and password, it will
    have full access to the API and an increased limit of 60 requests per minute.  Full
    access can increase data collection by 2x.

    Finally, for safety, it is recommended that the parameters below are not hard-coded
    directly into a program that uses Reddit Data Collector.  Rather, they should be
    kept in a separate credentials file as data which is then read into the program.
    (e.g. a JSON credentials file that is read into a program with a Python `with`
    clause).

    Parameters
    ----------
    client_id : str
        The client id for your Reddit application.

    client_secret : str
        The client secret for your Reddit application.

    user_agent : str
        A unique identifier that helps Reddit determine the souce of network requests.
        To use Reddit's API, you need a unique and descriptive user agent.  The
        following format is recommended:
            <platform>:<app ID>:<version string> (by u/<Reddit username>)

    username : str, default=None
        Your Reddit username.

    password : str, default=None
        Your Reddit password.

    Attributes
    ----------
    reddit : praw.Reddit
        An instance of the PRAW `Reddit` class that provides access to Reddit's API.

    Examples
    --------
    >>> import reddit_data_collector as rdc
    >>> # create instance of DataCollector
    >>> data_collector = rdc.DataCollector(
    ...     "<your_client_id>",
    ...     "<your_client_secret>",
    ...     "mac:script:v1.0 (by u/FakeRedditUser)",
    ...     "FakeRedditUser",
    ...     "FakePassword"
    ... )
    >>> # collect some data from Reddit
    >>> posts, comments = data_collector.get_data(
    ...     subreddits=["pics", "funny"],
    ...     post_filter="hot",
    ...     post_limit=10,
    ...     comment_data=True,
    ...     replies_data=True,
    ...     replace_more_limit=0
    ... )
    >>> # save data as .csv files
    >>> posts.to_csv("posts.csv", index=False)
    >>> comments.to_csv("posts.csv", index=False)
    """

    def __init__(
        self, client_id, client_secret, user_agent, username=None, password=None
    ):
        self.reddit = praw.Reddit(
            client_id=client_id,
            client_secret=client_secret,
            user_agent=user_agent,
            username=username,
            password=password,
        )

    def get_data(
        self,
        subreddits,
        post_filter="new",
        post_limit=None,
        top_post_filter=None,
        comment_data=True,
        replies_data=False,
        replace_more_limit=0,
        dataframe=True,
    ):
        """Collects post and comment data from Reddit.

        Parameters
        ----------
        subreddits : str or list of str
            The subreddit(s) to collect post and comment data from.

        post_filter : str, default="new"
            How to filter the subreddit posts.  Must be one of:  new, hot, or top.

        post_limit : int, default=None
            The number of posts to collect.  A limit of `None` sets the limit to
            the max allowed by Reddit's API, which is 1,000 in most cases.

        top_post_filter : str, default=None
            Determines how to filter the top posts for a subreddit.  Only required
            if `post_filter` is set to "top". Must be one of: all, day, hour, month,
            week, or year.

        comment_data : bool, default=True
            Whether or not to collect comment data for each post that is collected.
            If only post data is desired, set to `False`.  Only collecting posts can
            significantly speed up data collection since it will likely reduce the
            number of API requests by a lot.

        replies_data : bool, default=False
            Whether or not to collect the data for all replies to each comment that
            is collected for each post.  Setting this to `True` can cause the script
            to take arbitrarily long, as some reddit comments can have arbitrarily
            long reply threads.  Think carefully if you actually need this data before
            setting this parameter to True.  Often times, reply threads will contain
            useless data, since they often contain discussions of people trolling one
            another.

        replace_more_limit : int, default=0
            The number of PRAW `MoreComment` instances to replace when collecting
            comment data.  If you don't know what this means, the recommended
            setting is a value between 0 to 32.  A higher value means that
            potentially more comments will be collected in a sample. You can also
            set this to `None` which will ensure all comments and replies on a
            single post are collected.  Note that, setting this to an integer value
            higher than 32 or to `None` can significantly slow down the script,
            since this can increase the number of API calls drastically.

            For more info on the PRAW `MoreComment` class read this:
            https://praw.readthedocs.io/en/stable/tutorials/comments.html

        dataframe : bool, default=True
            Whether or not to return the collected data as a pandas DataFrame.
            If False, the data is returned in the raw form of a dictionary,
            where the keys for each dictionary are the subreddit name(s) and
            the values for each dictionary are the data collected.

        Returns
        -------
        posts, comments : pandas DataFrames
            The collected reddit data.
            If `comment_data` is False, `None` is returned for `comments`.

        See Also
        --------
        reddit_data_collector.io.to_pandas
            Used to convert raw `posts` or `comments` to a pandas `DataFrame`.
            Not needed if dataframe argument is left as True.

        reddit_data_collector.io.update_data
            Used to update an existing `.csv` file that contains prior data collected
            with Reddit Data Collector with new data collected.
        """
        if isinstance(subreddits, str):
            subreddits = [subreddits]

        self._verify_subreddits(subreddits)
        self._verify_post_filter(post_filter)

        if top_post_filter is not None:
            self._verify_top_post_filter(top_post_filter)

        posts = self._get_posts(subreddits, post_filter, post_limit, top_post_filter)

        if comment_data:
            comments = self._get_comments(posts, replies_data, replace_more_limit)
        else:
            comments = None

        if dataframe:
            posts = to_pandas(posts)

            if comments is not None:
                comments = to_pandas(comments)

        return posts, comments

    # ------------------------------HELPER FUNCTIONS------------------------------ #

    def _verify_subreddits(self, subreddits):
        """Verifies that each subreddit in a list of subreddits exist."""
        for subreddit in subreddits:
            if not self._check_subreddit_exists(subreddit):
                msg = f"r/{subreddit} does not exist"
                raise (SubredditError(msg))

    def _check_subreddit_exists(self, subreddit):
        """Checks if a subreddit exists."""
        # PRAW Subreddits instance
        subreddits = self.reddit.subreddits

        # may return numerous similar subreddits, first value should match
        exists = subreddits.search_by_name(subreddit)

        if not exists:
            return False
        else:
            return exists[0].display_name.lower() == subreddit.lower()

    def _verify_post_filter(self, post_filter):
        """Verifies that a post filter is valid.

        Raises FilterError if a invalid post filter is used.
        """
        if post_filter.lower() not in ["new", "hot", "top"]:
            msg = f"Invalid post_filter used: {post_filter}"
            raise (FilterError(msg))

    def _verify_top_post_filter(self, top_post_filter):
        """Verifies that a top post filter is valid.

        Raises FilterError if a invalid top post filter is used.
        """
        if top_post_filter.lower() not in [
            None,
            "all",
            "day",
            "hour",
            "month",
            "week",
            "year",
        ]:
            msg = f"Invalid top_post_filter used: {top_post_filter}"
            raise (FilterError(msg))

    def _get_posts(self, subreddits, post_filter, post_limit, top_post_filter):
        """Collects the post data for each subreddit in a list of subreddits."""
        posts = dict()

        for subreddit in subreddits:
            posts[subreddit] = self._get_subreddit_posts(
                subreddit, post_filter, post_limit, top_post_filter
            )

        return posts

    def _get_subreddit_posts(self, subreddit, post_filter, post_limit, top_post_filter):
        """Collects the post data for a single subreddit."""
        subreddit_posts = []

        # convert to PRAW Subreddit instance
        subreddit = self.reddit.subreddit(subreddit)

        desc = f"Collecting {post_filter} r/{subreddit} posts"

        # a "submission" is an instance of the PRAW Subission class
        if post_filter.lower() == "new":
            for submission in tqdm(subreddit.new(limit=post_limit), desc, post_limit):
                subreddit_posts.append(self._get_post_data(submission))

        elif post_filter.lower() == "hot":
            for submission in tqdm(subreddit.hot(limit=post_limit), desc, post_limit):
                subreddit_posts.append(self._get_post_data(submission))

        elif post_filter.lower() == "top":
            for submission in tqdm(subreddit.top(time_filter=top_post_filter), desc):
                subreddit_posts.append(self._get_post_data(submission))

        return subreddit_posts

    def _get_post_data(self, submission):
        """Collects the data for a single post in a subreddit."""
        post_data = {
            "subreddit_name": submission.subreddit.display_name,
            "post_created_utc": submission.created_utc,
            "id": submission.id,
            "is_original_content": submission.is_original_content,
            "is_self": submission.is_self,
            "link_flair_text": submission.link_flair_text,
            "locked": submission.locked,
            "num_comments": submission.num_comments,
            "over_18": submission.over_18,
            "score": submission.score,
            "spoiler": submission.spoiler,
            "stickied": submission.stickied,
            "title": submission.title,
            "upvote_ratio": submission.upvote_ratio,
            "url": submission.url,
        }

        return post_data

    def _get_comments(self, posts, replies_data, replace_more_limit):
        """Collects the comment data for each subreddit in a list of subreddits."""
        comments = dict()

        for subreddit, subreddit_post_data in posts.items():
            comments[subreddit] = self._get_subreddit_comments(
                subreddit, subreddit_post_data, replies_data, replace_more_limit
            )

        return comments

    def _get_subreddit_comments(
        self, subreddit, subreddit_post_data, replies_data, replace_more_limit
    ):
        """Collects the comment data for posts in a single subreddit."""
        subreddit_comments = []

        desc = f"Collecting comments for {len(subreddit_post_data)} r/{subreddit} posts"

        # a "submission" is an instance of the PRAW Subission class
        for post in tqdm(subreddit_post_data, desc, len(subreddit_post_data)):
            submission = self.reddit.submission(id=post["id"])
            submission.comments.replace_more(limit=replace_more_limit)

            if replies_data:
                for comment in submission.comments.list():
                    comment_data = self._get_comment_data(subreddit, comment)
                    subreddit_comments.append(comment_data)
            else:
                for comment in submission.comments:
                    comment_data = self._get_comment_data(subreddit, comment)
                    subreddit_comments.append(comment_data)

        return subreddit_comments

    def _get_comment_data(self, subreddit, comment):
        """Collects the data for a single comment on a subreddit post."""
        comment_data = {
            "subreddit_name": subreddit,
            "id": comment.id,
            "post_id": comment.link_id,
            "parent_id": comment.parent_id,
            "top_level_comment": comment.parent_id == comment.link_id,
            "body": comment.body,
            "comment_created_utc": comment.created_utc,
            "is_submitter": comment.is_submitter,
            "score": comment.score,
            "stickied": comment.stickied,
        }

        return comment_data
